language/vm/src/deserializer.rs

```diff
@@ -1510,3 +1517,214 @@ impl Opcodes {
+            | Bytecode::PackGeneric(_)
+            | Bytecode::Unpack(_)
+            | Bytecode::UnpackGeneric(_)
+            | Bytecode::ReadRef
+            | Bytecode::WriteRef
+            | Bytecode::FreezeRef
+            | Bytecode::MutBorrowField(_)
+            | Bytecode::MutBorrowFieldGeneric(_)
+            | Bytecode::ImmBorrowField(_)
+            | Bytecode::ImmBorrowFieldGeneric(_)
+            | Bytecode::MutBorrowGlobal(_)
+            | Bytecode::MutBorrowGlobalGeneric(_)
+            | Bytecode::ImmBorrowGlobal(_)
+            | Bytecode::ImmBorrowGlobalGeneric(_)
+            | Bytecode::Add
+            | Bytecode::Sub
+            | Bytecode::Mul
+            | Bytecode::Mod
+            | Bytecode::Div
+            | Bytecode::BitOr
+            | Bytecode::BitAnd
+            | Bytecode::Xor
+            | Bytecode::Or
+            | Bytecode::And
+            | Bytecode::Not
+            | Bytecode::Eq
+            | Bytecode::Neq
+            | Bytecode::Lt
+            | Bytecode::Gt
+            | Bytecode::Le
+            | Bytecode::Ge
+            | Bytecode::Abort
+            | Bytecode::Nop
+            | Bytecode::Exists(_)
+            | Bytecode::ExistsGeneric(_)
+            | Bytecode::MoveFrom(_)
+            | Bytecode::MoveFromGeneric(_)
+            | Bytecode::MoveTo(_)
+            | Bytecode::MoveToGeneric(_)
+            | Bytecode::Shl
+            | Bytecode::Shr => (),
+        }
+    }
+    script.code.code = {
+        let old_code = std::mem::take(&mut script.code.code);
+        borrow_owned_signers_instructions.extend(old_code);
+        borrow_owned_signers_instructions
+    };
+
+    // Build a mapping of the signatures to their old index
+    // Update signatures
+    let mut signature_to_index = script
+        .signatures
+        .iter()
+        .enumerate()
+        .map(|(idx, sig)| (sig.clone(), idx))
+        .collect::<BTreeMap<_, _>>();
+    let new_parameters_idx = match signature_to_index.get(&new_parameters) {
+        Some(idx) => *idx,
+        None => {
+            script.signatures.push(new_parameters.clone());
+            let idx = signature_to_index.len();
+            signature_to_index.insert(new_parameters, idx);
+            idx
+        }
+    };
+    let new_locals_signature_idx = match signature_to_index.get(&new_locals_signature) {
+        Some(idx) => *idx,
+        None => {
+            script.signatures.push(new_locals_signature.clone());
+            let idx = signature_to_index.len();
+            signature_to_index.insert(new_locals_signature, idx);
+            idx
+        }
+    };
+    let table_index_max = TableIndex::MAX as usize;
+    if new_parameters_idx > table_index_max || new_locals_signature_idx > table_index_max {
+        // TODO
+        panic!()
+    }
+    script.parameters = SignatureIndex(new_parameters_idx as TableIndex);
+    script.code.locals = SignatureIndex(new_locals_signature_idx as TableIndex);
+}
```

Comment:
This conversion looks pretty scary to me TBH, and if I'm understanding correctly, it's also doing a lot of work. Would it be better if we do the following instead?
1. The bytecode verifier allows only `&signer` for v1 scripts and only `signer` for v2 scripts
2. Until we drop support for all legacy scripts, the VM will continue supporting both `&signer` and `signer`. This is safe because of the verifier check above.

Reply:
I would really rather not do that as it is not in line with our current migration plans. All of our current backwards compatibility models load the old version into the new one, and this is necessary to stay inline with that. 

Currently no other aspect of Move gates itself on the Version # (it is all contained inside the deserializer). This is intentional to keep things from getting out of hand.

Reply:
Although given that we are stuck with the signer ref business because of execute_function and backwards compat, this might not be that bad of an option... 

I'll think about it

Reply:
I share Victor's concern, although I'm conflicted because I really want to kill the `&signer` logic in the VM too...

- One other worry is that will change the gas cost of all the legacy transaction scripts: they'll have an extra borrow on the `signer`. But maybe the gas cost was already going to change because of the shim logic? In either case, we should announce the gas cost changes prominently in the changelog and other applicable channels (CC @tzakian )
- Another (extreme corner case) is that this could push an otherwise-legal bytecode program over the max locals or max instructions thresholds. Obviously, none of our Diem scripts look like this, but it's technically a violation of backward compatibility.

Reply:
This seems like a non-starter to me -- we need to execute legacy scripts with exactly the same gas consumption, right? I must have misunderstood because I thought this change would only apply to the new ScriptFunction transactions.